### ISSUE TITLE
Not Fixed: Inconsistent timeStampPeriod in vulkanCapsViewer for non-Apple GPUs

### ISSUE THREAD

**Issue by SRSaunders** [KhronosGroup/MoltenVK#2040]
When I run **vulkanCapsViewer** I get an inconsistent value for `timeStampPeriod` on non-Apple GPUs.  The value changes from one invocation to the next.  I suspect this may have something to do with the new low-pass filter.

I can solve this issue in my own applications by setting `mvkConfig.timestampPeriodLowPassAlpha = 1.0`.  However **vulkanCapsViewer** does not set this value and reports a filtered value for the timeStampPeriod which is likely incorrect.

**UPDATE**: I am wondering whether it would be beneficial to initialize MoltenVK's startup `_mvkTimestampPeriod` value with known nominal values by GPU vendor and platform arch.  That way the lowpass filter would not disturb the nominal value too much during initial (possibly noisy) startup sampling for applications that do not set the filter alpha value.  There are only a few GPU vendors/arch platforms for macOS and all have known nominal values: AMD/Intel = 1, Nvidia/Intel=1, Intel/Intel=? (not sure), Apple Silicon(macOS)=1.  Rosetta2 may also need to be factored in for Apple Silicon on macOS. Just an idea but something to consider.  Nominal initial values for lowpass filters can help achieve more rapid convergence.

See the following screen caps taken a few minutes apart (actual value should be closer to 1.0):

![Screenshot 2023-10-13 at 2 40 39 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/5d20eac6-865c-464f-9fa9-2f99a05ffb31)

![Screenshot 2023-10-13 at 2 46 30 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/48f4cd27-2c75-43ff-96ff-cae576c6695a)

![Screenshot 2023-10-13 at 3 05 10 PM](https://github.com/KhronosGroup/MoltenVK/assets/82544213/b3d11c63-ff4a-4b12-8b5a-384106160826)

**Comment by billhollings** [comment]
PR #2043 should fix this issue. Please retest with latest MoltenVK, and close this issue if it resolves the issue.

The default value of `MVKConfiguration::timestampPeriodLowPassAlpha` is now `1.0`, indicating that, on non-Apple GPUs, the latest timestamp calculation value will be used without any filter, by default.

BTW...to adjust the value of `timestampPeriodLowPassAlpha` on binary apps like `vulkanCapsViewer`, you can start the app with the environment variable `MVK_CONFIG_TIMESTAMP_PERIOD_LOWPASS_ALPHA` set to the value you want.

**Comment by SRSaunders** [comment]
This might be tricky to test using **vulkanCapsViewer** until a new SDK is released.  Doesn't this app use the Vulkan system installation and Vulkan loader?  I am not sure how to override with a custom build of MoltenVK.  I will retest once a new SDK is released containing MoltenVK 1.2.6

**Comment by richard-lunarg** [comment]
FWIW, I tried reproducing this with the last SDK and could not. This was on an Intel/NVidia Mac. The problem seems like it might be specific to AMD. Still... a new SDK should be available for you early next week for testing on your system.

**Comment by SRSaunders** [comment]
Thanks - I look forward to the new SDK and will test following release.

**Comment by SRSaunders** [comment]
I tested this again on SDK 1.3.268.1 and the problem remains for both **vulkanCapsViewer** and the command line **vulkaninfo** command, or for any application that calls `vkGetPhysicalDeviceProperties` (to get `timestampPeriod`) only a _**single time**_ when running on a non-Apple GPU.

I looked at the `updateTimestampPeriod()` code in MoltenVK and I now understand the issue:

The values of `_prevCPUTimestamp` and `_prevGPUTimestamp` are both initialized to 0 at startup.  For non-Apple GPUs, `updateTimestampPeriod()` then calculates the time stamp calibration using the previous time sample (0) and the current time sample when `vkGetPhysicalDeviceProperties` is called.  This leads to some incorrect random number on the first call.  For applications that call `vkGetPhysicalDeviceProperties` multiple times in sequence this is not a real problem as the second and subsequent calls return valid values for `timestampPeriod`.  However, applications like **vulkanCapsViewer** and **vulkaninfo**, and perhaps others, call `vkGetPhysicalDeviceProperties` only a single time and will get invalid values.

To correct this, I have prototyped the following change in `MVKPhysicalDevice::initMTLDevice` to initialize the `_prevCPUTimestamp` and `_prevGPUTimestamp` values at MoltenVK startup so the initial call to `updateTimestampPeriod()` returns correct values.  This seems to fix the problem:

```
...

_properties.limits.timestampComputeAndGraphics = VK_TRUE;
_properties.limits.timestampPeriod = 1.0;	// On non-Apple GPU's, this can vary over time, and is calculated based on actual GPU activity.

// Must initialize _prevCPUTimestamp and _prevGPUTimestamp otherwise initial call to updateTimestampPeriod() will return invalid results
if (_properties.vendorID != kAppleVendorId &&
	[_mtlDevice respondsToSelector: @selector(sampleTimestamps:gpuTimestamp:)]) {

	[_mtlDevice sampleTimestamps: &_prevCPUTimestamp gpuTimestamp: &_prevGPUTimestamp];
}
...
```
